crypto/cms/cms_sd.c

```diff
@@ -760,17 +765,22 @@ static int cms_SignerInfo_content_sign(CMS_ContentInfo *cms,
         unsigned char *sig;
         unsigned int siglen;
 
-        sig = OPENSSL_malloc(EVP_PKEY_get_size(si->pkey));
+        sig = OPENSSL_malloc(md != NULL ? mdlen : EVP_PKEY_get_size(si->pkey));
         if (sig == NULL) {
             ERR_raise(ERR_LIB_CMS, ERR_R_MALLOC_FAILURE);
             goto err;
         }
-        if (!EVP_SignFinal_ex(mctx, sig, &siglen, si->pkey,
-                              ossl_cms_ctx_get0_libctx(ctx),
-                              ossl_cms_ctx_get0_propq(ctx))) {
-            ERR_raise(ERR_LIB_CMS, CMS_R_SIGNFINAL_ERROR);
-            OPENSSL_free(sig);
-            goto err;
+        if (md != NULL) {
+            memcpy(sig, md, mdlen);
+            siglen = mdlen;
+        } else {
+            if (!EVP_SignFinal_ex(mctx, sig, &siglen, si->pkey,
+                                  ossl_cms_ctx_get0_libctx(ctx),
+                                  ossl_cms_ctx_get0_propq(ctx))) {
+                ERR_raise(ERR_LIB_CMS, CMS_R_SIGNFINAL_ERROR);
+                OPENSSL_free(sig);
+                goto err;
+            }
```

Comment:
Actually using supplied hash in this case does not make much sense to me at all. Because here we are just finishing the signing (there were possibly some EVP_SignInit/EVP_SignUpdate calls before).  There is not much you could do with the precomputed digest here.  So there should be an error reported if md is non NULL in this case.

Add a new CMS_R_OPERATION_UNSUPPORTED or something like this. (Just add ERR_raise call and do `make update` to add the reason code to the appropriate header etc.)

Reply:
Done. Thanks for your help.